tests/bsim/bluetooth/host/l2cap/multilink_peripheral/src/dut.c

```diff
@@ -0,0 +1,337 @@
+};
+
+static struct test_ctx contexts[CONFIG_BT_MAX_CONN];
+
+static int send_data_over_l2cap(struct bt_l2cap_chan *chan, uint8_t *data, size_t len)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(chan->conn), addr, sizeof(addr));
+
+	LOG_DBG("[%s] chan %p data %p len %d", addr, chan, data, len);
+
+	struct net_buf *buf = net_buf_alloc(&sdu_tx_pool, K_NO_WAIT);
+
+	if (buf == NULL) {
+		TEST_FAIL("No more memory");
+		return -ENOMEM;
+	}
+
+	net_buf_reserve(buf, BT_L2CAP_SDU_CHAN_SEND_RESERVE);
+	net_buf_add_mem(buf, data, len);
+
+	int ret = bt_l2cap_chan_send(chan, buf);
+
+	TEST_ASSERT(ret == 0, "Failed sending: err %d", ret);
+	LOG_DBG("queued SDU", len);
+
+	return ret;
+}
+
+static void continue_sending(struct test_ctx *ctx)
+{
+	struct bt_l2cap_chan *chan = &ctx->le_chan.chan;
+
+	TEST_ASSERT(ctx->le_chan.state == BT_L2CAP_CONNECTED, "such race much wow");
+
+	LOG_DBG("%p, transmitted %d SDUs", chan, ctx->sdu_count);
+
+	if (ctx->sdu_count < SDU_NUM) {
+		send_data_over_l2cap(chan, tx_data, sizeof(tx_data));
+	} else {
+		char addr[BT_ADDR_LE_STR_LEN];
+
+		bt_addr_le_to_str(bt_conn_get_dst(chan->conn), addr, sizeof(addr));
+
+		LOG_DBG("[%s] Done sending", addr);
+	}
+}
+
+static struct test_ctx *get_ctx_from_chan(struct bt_l2cap_chan *chan)
+{
+	struct bt_l2cap_le_chan *le_chan = CONTAINER_OF(chan, struct bt_l2cap_le_chan, chan);
+	struct test_ctx *ctx = CONTAINER_OF(le_chan, struct test_ctx, le_chan);
+
+	TEST_ASSERT(PART_OF_ARRAY(contexts, ctx), "memory corruption");
+
+	return ctx;
+}
+
+static void sent_cb(struct bt_l2cap_chan *chan)
+{
+	struct test_ctx *ctx = get_ctx_from_chan(chan);
+
+	LOG_DBG("%p", chan);
+
+	ctx->sdu_count++;
+
+	continue_sending(ctx);
+}
+
+static void l2cap_chan_connected_cb(struct bt_l2cap_chan *chan)
+{
+	struct bt_l2cap_le_chan *le_chan = CONTAINER_OF(chan, struct bt_l2cap_le_chan, chan);
+
+	LOG_DBG("%p (tx mtu %d mps %d) (tx mtu %d mps %d)", chan, le_chan->tx.mtu, le_chan->tx.mps,
+		le_chan->rx.mtu, le_chan->rx.mps);
+
+	LOG_DBG("initiating SDU transfer");
+	continue_sending(get_ctx_from_chan(chan));
+}
+
+static void l2cap_chan_disconnected_cb(struct bt_l2cap_chan *chan)
+{
+	LOG_DBG("%p", chan);
+}
+
+static int recv_cb(struct bt_l2cap_chan *chan, struct net_buf *buf)
+{
+	TEST_FAIL("DUT should not receive data");
+
+	return 0;
+}
+
+static struct bt_l2cap_chan_ops ops = {
+	.connected = l2cap_chan_connected_cb,
+	.disconnected = l2cap_chan_disconnected_cb,
+	.recv = recv_cb,
+	.sent = sent_cb,
+};
+
+static int connect_l2cap_channel(struct bt_conn *conn, struct bt_l2cap_le_chan *le_chan)
+{
+	memset(le_chan, 0, sizeof(*le_chan));
+	le_chan->chan.ops = &ops;
+
+	return bt_l2cap_chan_connect(conn, &le_chan->chan, 0x0080);
+}
+
+static bool addr_in_use(bt_addr_le_t *address)
+{
+	return !bt_addr_le_eq(address, BT_ADDR_LE_ANY);
+}
+
+static struct test_ctx *alloc_ctx(void)
+{
+	for (int i = 0; i < ARRAY_SIZE(contexts); i++) {
```

Comment:
```suggestion
	for (size_t i = 0; i < ARRAY_SIZE(contexts); i++) {
```

Minor suggestion to use the same type for the iterator and the compared value